subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultResolvableDependenciesConfiguration.java

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright 2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.artifacts.configurations;
+
+import org.gradle.api.artifacts.ConfigurablePublishArtifact;
+import org.gradle.api.artifacts.DependenciesConfiguration;
+import org.gradle.api.artifacts.DependencyResolutionListener;
+import org.gradle.api.artifacts.ResolvableConfiguration;
+import org.gradle.api.capabilities.Capability;
+import org.gradle.api.internal.DomainObjectContext;
+import org.gradle.api.internal.artifacts.ConfigurationResolver;
+import org.gradle.api.internal.artifacts.ResolveExceptionContextualizer;
+import org.gradle.api.internal.artifacts.component.ComponentIdentifierFactory;
+import org.gradle.api.internal.artifacts.dsl.dependencies.DependencyLockingProvider;
+import org.gradle.api.internal.artifacts.ivyservice.moduleconverter.RootComponentMetadataBuilder;
+import org.gradle.api.internal.attributes.ImmutableAttributesFactory;
+import org.gradle.api.internal.collections.DomainObjectCollectionFactory;
+import org.gradle.api.internal.file.FileCollectionFactory;
+import org.gradle.api.internal.project.ProjectStateRegistry;
+import org.gradle.api.internal.tasks.TaskDependencyFactory;
+import org.gradle.configuration.internal.UserCodeApplicationContext;
+import org.gradle.internal.Factory;
+import org.gradle.internal.event.ListenerBroadcast;
+import org.gradle.internal.model.CalculatedValueContainerFactory;
+import org.gradle.internal.operations.BuildOperationExecutor;
+import org.gradle.internal.reflect.Instantiator;
+import org.gradle.internal.typeconversion.NotationParser;
+import org.gradle.internal.work.WorkerThreadRegistry;
+
+/**
+ * Extends a {@link DefaultConfiguration} to expose it as a {@link ResolvableConfiguration} and a {@link DependenciesConfiguration}.
+ */
+public class DefaultResolvableDependenciesConfiguration extends DefaultConfiguration implements ResolvableConfiguration, DependenciesConfiguration {
```

Comment:
I do not think making these pre-wired factory method things are going to be quite that easy. Writing the factory may be easy, but I think you're underestimating the difficulty of migrating to use it. 

There are many of these resolvable + dependencies configurations in the Gradle codebase currently -- and I listed all of them in the spec. However, to split these up into two separate configurations, we need to pick a new name for the new one. We cannot have both a resolvable and a dependencies configuration in the `ConfigurationContainer` with the same name. If we chose a different name for the new configuration, as we would need to, that would be a breaking change. 

While it may be goal to remove these types of configurations, they are very prevalent. Splitting up these configurations is not in the scope of this spec or PR. Making these kinds of changes would be quite a large undertaking. Not saying we shouldn't eventually do it, but for the time being this PR needs to support the existing configurations which are already in this state.

Reply:
I would argue locked configuration roles is much more valuable than 1 role per configuration. In practice, what technical issues does having a resolvable + dependencies configuration cause? We do not have a problem building dependency management metadata for these mixed configurations. 

I think we need to discuss this more -- there is definitely community opinions which are also a factor in and not ignore -- see: https://docs.google.com/document/d/1a2vtM10FiWdTpnY8b2S-q28Cl0xUEVIzAOl3BpOeyng/edit?disco=AAAAsiOEGg4

In any case we have 18 of these kinds of configurations. This is not something we can ignore right now. We can make efforts to remove them if we deem that necessary, though in order to support the existing first-party configurations without falling back to mutating configuration roles after creation, factory methods that create resolvable + dependencies configuration are required.

Reply:
As long as configurations can be multi-role, we can't set up separate role-based containers for them.  We also have (more/different) conceptual problems IRT inheritance via `extendsFrom`.  How do we handle timing issues - for example: what if I try to add a dependency after this configuration has been resolved - doesn't it have to say no, which makes the behavior inconsistent with pure declarables?  I'd also argue that the growth of methods and interfaces needed to support them is itself a problem.

Why try and solve these issues now?  Let multi-role configurations remain legacy configurations for this change, which is big enough already, and only add new factories for single-role configurations.  The community continues to use legacy configurations and we think a little more on the need for this.

Reply:
> The community continues to use legacy configurations

The resolvable + dependencies factory methods are internal

Reply:
Right, I'm saying we have to continue to support them during this change, so providing a different internal API doesn't gain us much.

Reply:
It gains us the ability to deprecate calling role-mutating methods. If we rely on those methods to turn a legacy configuring into a Resolvable + dependencies configuration, we can't deprecate those methods

Reply:
Why couldn't we just suppress the deprecation warning where we're forced to call them?

Reply:
We could do that. Though we'd be doing that the same way for ~16 configurations so refactoring that into a method is a logical step here. 

I feel like we're arguing over peanuts here. The `DefaultResolvableDependenciesConfiguration` type has been removed in the most recent version of this PR. Both a "legacy" configuration, created with `create` and `register` and `resolvableDependenciesUnlocked` configurations are backed by the same underlying type here, `DefaultUnlockedConfiguration` and only differ by the role we pass in at creation. 

By providing a factory method for a configuration type we use often, whether we choose to continue using this configuration type in the future, we gain the following advantages:
1. `roleAtCreation` reflects the actual intended role of the configuration
2. We avoid using methods we are intending to avoid (mutating role methods)
3. We avoid using `DeprecationLogger.whileDisabled` which IMO is very smelly
4. We make it easy to discover which configurations are resolvable + dependencies by allowing IDE navigation

In the current state of the PR, we do not call `setCanBeX` anywhere (with one small edge case -- `JvmPluginServices#registerDependencyBucket` -- which is unused and can also remove in this PR). This is a huge win IMO

Reply:
I added archUnit rules to make sure we don't make more of these types of configurations